com.unity.shadergraph/Editor/Generation/Processors/Generator.cs

```diff
@@ -167,11 +167,14 @@ void BuildShader()
                 Debug.LogError($"Error in Shader Graph {graphName}:{ShaderKeyword.kVariantLimitWarning}");
 
                 m_ConfiguredTextures = shaderProperties.GetConfiguredTextures();
-                m_Builder.AppendLines(ShaderGraphImporter.k_ErrorShader);
+                m_Builder.AppendLines(ShaderGraphImporter.k_ErrorShader.Replace("Hidden/GraphErrorShader2", graphName));
             }
 
             foreach (var activeNode in activeNodeList.OfType<AbstractMaterialNode>())
+            {
+                activeNode.SetUsedByGenerator();
                 activeNode.CollectShaderProperties(shaderProperties, m_Mode);
+            }
```

Comment:
At this point, the generator has produced the list of "actually used nodes" in "activeNodeList", so we just push that down to the nodes themselves as a flag.